File: samples/facade_sample.js
/**
 * Sample implementations of the Facade pattern in JavaScript.
 * 
 * The Facade pattern provides a simplified interface to a complex subsystem.
 * It doesn't encapsulate the subsystem but provides a unified interface to make
 * the subsystem easier to use.
 */

// Implementation 1: Classic Facade Pattern

/**
 * A complex subsystem component
 */
class ComplexSubsystemA {
  operation1() {
    return "Subsystem A: Operation 1";
  }
  
  operation2() {
    return "Subsystem A: Operation 2";
  }
}

/**
 * Another complex subsystem component
 */
class ComplexSubsystemB {
  operation1() {
    return "Subsystem B: Operation 1";
  }
  
  operation2() {
    return "Subsystem B: Operation 2";
  }
}

/**
 * Yet another complex subsystem component
 */
class ComplexSubsystemC {
  operation1() {
    return "Subsystem C: Operation 1";
  }
}

/**
 * The Facade class provides a simple interface to the complex subsystem
 */
class Facade {
  constructor() {
    this.subsystemA = new ComplexSubsystemA();
    this.subsystemB = new ComplexSubsystemB();
    this.subsystemC = new ComplexSubsystemC();
  }
  
  /**
   * The facade methods that provide a simplified interface.
   * They delegate to subsystem objects to perform complex operations.
   */
  operation() {
    const results = [];
    results.push("Facade initializes subsystems:");
    results.push(this.subsystemA.operation1());
    results.push(this.subsystemB.operation1());
    results.push("Facade orders subsystems to perform the action:");
    results.push(this.subsystemA.operation2());
    results.push(this.subsystemB.operation2());
    results.push(this.subsystemC.operation1());
    return results.join('\n');
  }
}

// Implementation 2: Module Facade Pattern with ES6 Modules

/**
 * API service for external HTTP requests
 */
class APIService {
  constructor(baseURL) {
    this.baseURL = baseURL;
  }
  
  get(endpoint) {
    return `GET ${this.baseURL}/${endpoint}`;
  }
  
  post(endpoint, data) {
    return `POST ${this.baseURL}/${endpoint} with ${JSON.stringify(data)}`;
  }
  
  put(endpoint, data) {
    return `PUT ${this.baseURL}/${endpoint} with ${JSON.stringify(data)}`;
  }
  
  delete(endpoint) {
    return `DELETE ${this.baseURL}/${endpoint}`;
  }
}

/**
 * Authentication service
 */
class AuthService {
  constructor() {
    this.token = null;
  }
  
  login(credentials) {
    this.token = "sample-jwt-token";
    return `Logged in as ${credentials.username}`;
  }
  
  logout() {
    this.token = null;
    return "Logged out";
  }
  
  getToken() {
    return this.token;
  }
}

/**
 * Storage service for local data persistence
 */
class StorageService {
  setItem(key, value) {
    return `Stored ${key}=${value}`;
  }
  
  getItem(key) {
    return `Retrieved ${key}`;
  }
  
  removeItem(key) {
    return `Removed ${key}`;
  }
  
  clear() {
    return "Storage cleared";
  }
}

/**
 * User service facade that simplifies working with users through multiple services
 */
class UserServiceFacade {
  constructor(apiBaseURL) {
    this.api = new APIService(apiBaseURL);
    this.auth = new AuthService();
    this.storage = new StorageService();
  }
  
  /**
   * Login a user and store their information
   */
  login(username, password) {
    // Authenticate
    const result = this.auth.login({ username, password });
    
    // Store token
    this.storage.setItem('auth_token', this.auth.getToken());
    
    // Load user data
    const userData = this.api.get(`users/${username}`);
    
    // Store user data
    this.storage.setItem('user_data', userData);
    
    return `User ${username} logged in successfully`;
  }
  
  /**
   * Register a new user
   */
  register(userData) {
    // Create user account
    const result = this.api.post('users', userData);
    
    // Store registration status
    this.storage.setItem('registration_status', 'complete');
    
    return `User ${userData.username} registered successfully`;
  }
  
  /**
   * Logout the current user
   */
  logout() {
    // Logout from auth service
    this.auth.logout();
    
    // Clear stored data
    this.storage.removeItem('auth_token');
    this.storage.removeItem('user_data');
    
    return "User logged out successfully";
  }
}

// Implementation 3: Object Literal Facade Pattern

/**
 * These would normally be more complex and possibly in separate files
 */
// Audio handling
const audioEngine = {
  init() {
    return "Audio engine initialized";
  },
  playSound(soundId) {
    return `Playing sound: ${soundId}`;
  },
  stopSound(soundId) {
    return `Stopping sound: ${soundId}`;
  },
  setVolume(level) {
    return `Setting volume to ${level}`;
  }
};

// Graphics handling
const graphicsEngine = {
  init(width, height) {
    return `Graphics initialized with resolution ${width}x${height}`;
  },
  drawSprite(spriteId, x, y) {
    return `Drawing sprite ${spriteId} at position (${x}, ${y})`;
  },
  clearScreen() {
    return "Screen cleared";
  }
};

// Input handling
const inputHandler = {
  init() {
    return "Input handler initialized";
  },
  bindKey(key, action) {
    return `Key ${key} bound to action: ${action}`;
  },
  unbindKey(key) {
    return `Key ${key} unbound`;
  }
};

/**
 * Game facade using object literal that simplifies working with complex subsystems
 */
const gameFacade = {
  // Initialize all subsystems
  init(width, height) {
    const results = [];
    results.push(audioEngine.init());
    results.push(graphicsEngine.init(width, height));
    results.push(inputHandler.init());
    
    // Bind default keys
    results.push(inputHandler.bindKey("SPACE", "JUMP"));
    results.push(inputHandler.bindKey("W", "MOVE_UP"));
    results.push(inputHandler.bindKey("A", "MOVE_LEFT"));
    
    return results.join('\n');
  },
  
  // Start a new game level
  startLevel(levelId) {
    const results = [];
    results.push(graphicsEngine.clearScreen());
    results.push(`Loading level ${levelId}`);
    results.push(audioEngine.playSound("LEVEL_START"));
    return results.join('\n');
  },
  
  // Handle player jumping
  jump() {
    const results = [];
    results.push(audioEngine.playSound("JUMP"));
    results.push(graphicsEngine.drawSprite("PLAYER_JUMP", 100, 150));
    return results.join('\n');
  }
};

// Usage examples
function runExamples() {
  // Example 1: Classic Facade
  console.log("Example 1: Classic Facade");
  const facade = new Facade();
  console.log(facade.operation());
  
  console.log("-".repeat(50));
  
  // Example 2: Module Facade
  console.log("Example 2: Module Facade");
  const userService = new UserServiceFacade("https://api.example.com");
  console.log(userService.register({ username: "john_doe", email: "john@example.com" }));
  console.log(userService.login("john_doe", "password123"));
  console.log(userService.logout());
  
  console.log("-".repeat(50));
  
  // Example 3: Object Literal Facade
  console.log("Example 3: Object Literal Facade");
  console.log(gameFacade.init(1920, 1080));
  console.log(gameFacade.startLevel(1));
  console.log(gameFacade.jump());
}

// Uncomment to run the examples
// runExamples();

File: samples/facade_sample.py
"""
Sample implementations of the Facade pattern in Python.

The Facade pattern provides a simplified interface to a complex subsystem.
It doesn't encapsulate the subsystem but provides a unified interface to make
the subsystem easier to use.
"""

# Implementation 1: Classic Facade Pattern


class ComplexSubsystemA:
    """A complex subsystem component."""
    
    def operation1(self) -> str:
        return "Subsystem A: Operation 1"
    
    def operation2(self) -> str:
        return "Subsystem A: Operation 2"


class ComplexSubsystemB:
    """Another complex subsystem component."""
    
    def operation1(self) -> str:
        return "Subsystem B: Operation 1"
    
    def operation2(self) -> str:
        return "Subsystem B: Operation 2"


class ComplexSubsystemC:
    """Yet another complex subsystem component."""
    
    def operation1(self) -> str:
        return "Subsystem C: Operation 1"


class Facade:
    """Facade provides a simple interface to the complex subsystem."""
    
    def __init__(self):
        """Initialize the facade with subsystem objects."""
        self._subsystem_a = ComplexSubsystemA()
        self._subsystem_b = ComplexSubsystemB()
        self._subsystem_c = ComplexSubsystemC()
    
    def operation(self) -> str:
        """
        The facade method that provides a simplified interface.
        It delegates to subsystem objects to perform complex operations.
        """
        results = []
        results.append("Facade initializes subsystems:")
        results.append(self._subsystem_a.operation1())
        results.append(self._subsystem_b.operation1())
        results.append("Facade orders subsystems to perform the action:")
        results.append(self._subsystem_a.operation2())
        results.append(self._subsystem_b.operation2())
        results.append(self._subsystem_c.operation1())
        return "\n".join(results)


# Implementation 2: Facade with Configuration


class DatabaseService:
    """Database service for complex database operations."""
    
    def connect(self, connection_string):
        return f"Connected to database with {connection_string}"
    
    def execute_query(self, query):
        return f"Executing query: {query}"
    
    def close_connection(self):
        return "Database connection closed"


class CacheService:
    """Cache service for handling caching operations."""
    
    def get(self, key):
        return f"Getting {key} from cache"
    
    def set(self, key, value):
        return f"Setting {key}={value} in cache"
    
    def clear(self):
        return "Clearing cache"


class LogService:
    """Logging service for various log levels."""
    
    def info(self, message):
        return f"INFO: {message}"
    
    def error(self, message):
        return f"ERROR: {message}"
    
    def debug(self, message):
        return f"DEBUG: {message}"


class DataServiceFacade:
    """
    Facade that provides a simplified interface to database, cache, and logging services.
    """
    
    def __init__(self, db_connection_string):
        """Initialize the facade with the provided configuration."""
        self.db = DatabaseService()
        self.cache = CacheService()
        self.logger = LogService()
        self.connection_string = db_connection_string
    
    def get_data(self, key, query):
        """Get data first checking cache, then falling back to database query."""
        # Log operation
        self.logger.info(f"Getting data for {key}")
        
        # Try from cache first
        cache_result = self.cache.get(key)
        self.logger.debug(f"Cache result: {cache_result}")
        
        # If not in cache, get from database
        self.logger.info("Getting data from database")
        self.db.connect(self.connection_string)
        db_result = self.db.execute_query(query)
        self.db.close_connection()
        
        # Update cache
        self.cache.set(key, "result_from_query")
        
        return f"Data retrieved for {key}"
    
    def clear_cache(self):
        """Clear the cache and log the operation."""
        self.logger.info("Clearing all cache")
        result = self.cache.clear()
        return result


# Implementation 3: Module-level Facade


# These would normally be in separate modules
def _complex_algorithm_part1(data):
    return f"Processing {data} with algorithm part 1"


def _complex_algorithm_part2(data):
    return f"Processing {data} with algorithm part 2"


def _format_result(result):
    return f"Formatted: {result}"


# Facade function that simplifies the algorithm usage
def process_data(data):
    """
    A function facade that simplifies the usage of a complex algorithm.
    """
    # Step 1: Run first part of algorithm
    intermediate_result = _complex_algorithm_part1(data)
    
    # Step 2: Run second part of algorithm
    final_result = _complex_algorithm_part2(intermediate_result)
    
    # Step 3: Format the result
    formatted_result = _format_result(final_result)
    
    return formatted_result


# Usage Example
if __name__ == "__main__":
    # Example 1: Classic Facade
    print("Example 1: Classic Facade")
    facade = Facade()
    result = facade.operation()
    print(result)
    
    print("-" * 50)
    
    # Example 2: Configurable Facade
    print("Example 2: Configurable Facade")
    data_service = DataServiceFacade("mysql://localhost:3306/mydatabase")
    result = data_service.get_data("user:123", "SELECT * FROM users WHERE id=123")
    print(result)
    result = data_service.clear_cache()
    print(result)
    
    print("-" * 50)
    
    # Example 3: Function Facade
    print("Example 3: Function Facade")
    result = process_data("raw_data")
    print(result)